addons/portal/controllers/thread.py

```diff
@@ -0,0 +1,21 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request
+from odoo.addons.mail.controllers import thread
+from odoo.addons.portal.utils import validate_thread_with_hash_pid, validate_thread_with_token
+
+
+class ThreadController(thread.ThreadController):
+
+    def _prepare_post_data(self, post_data, thread, **kwargs):
+        post_data = super()._prepare_post_data(post_data, thread, **kwargs)
+        if request.env.user._is_public():
+            if validate_thread_with_hash_pid(thread, kwargs.get("hash"), kwargs.get("pid")):
+                post_data["author_id"] = int(kwargs["pid"])
+            elif (
+                hasattr(thread, "partner_id")
```

Comment:
Probably time to use ``_mail_get_partner_fields`` method that returns the customer field (allowing to be less hardcoded, notably if you have invoice_partner_id for example).